rhine-terminal/tests/Main.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE NamedFieldPuns #-}
+module Main where
+
+-- base
+import Prelude hiding (putChar)
+import GHC.Conc (retry, readTVarIO, atomically)
+import Control.Concurrent (forkIO, threadDelay)
+import Control.Monad (void)
+
+-- rhine
+import FRP.Rhine
+import FRP.Rhine.Terminal
+
+-- terminal
+import System.Terminal
+import System.Terminal.Internal
+
+-- text
+import Data.Text (singleton)
+
+-- stm
+import Control.Concurrent.STM.TQueue
+
+-- hspec
+import Test.Hspec
+
+type KeyClock = SelectClock TerminalEventClock Char
+
+keyClock :: KeyClock
+keyClock = SelectClock { mainClock = TerminalEventClock , select }
+  where
+    select :: Tag TerminalEventClock -> Maybe Char
+    select (Right (KeyEvent (CharKey k) _)) = Just k
+    select _                                = Nothing
+
+defaultSettings :: TQueue Event -> VirtualTerminalSettings
+defaultSettings eventQueue = VirtualTerminalSettings
+  { virtualType         = "xterm"
+  , virtualWindowSize   = pure (Size 3 10)
+  , virtualEvent        = readTQueue eventQueue
+  , virtualInterrupt    = retry
+  }
+
+displayDot :: MonadScreen m => ClSF m KeyClock () ()
+displayDot = constMCl $ do
+  putChar '.'
+  flush
+
+testRhine :: Terminal t => Rhine (TerminalT t IO) KeyClock () ()
+testRhine = displayDot @@ keyClock
+
+charEvent :: Terminal t => TQueue Event -> t -> Char -> IO ()
+charEvent eventQueue terminal char = do
+  termCommand terminal $ PutText $ singleton char
+  atomically $ writeTQueue eventQueue $ KeyEvent (CharKey char) mempty
+
+main :: IO ()
+main = hspec $ do
+  describe "rhine-terminal with VirtualTerminal" $ do
+    it "replaces virtual inputs by dots" $ do
```

Comment:
This description isn't quite true anymore.

Reply:
Oh, indeed I've changed the test to go back to the original idea.